internal/docstore/dynamodocstore/dynamo.go

```diff
@@ -279,6 +283,75 @@ func (c *collection) primaryKey(a *driver.Action) [2]interface{} {
 	return keys
 }
 
+func (c *collection) runActionsUnordered(ctx context.Context, actions []*driver.Action, opts *driver.RunActionsOptions) driver.ActionListError {
+	errs := make([]error, len(actions))
+	groups, err := c.splitActionsUnordered(actions)
+	if err != nil {
+		errs[0] = err
+		return driver.NewActionListError(errs)
+	}
+	var wg sync.WaitGroup
+	groupBaseIndex := 0 // index in actions of first action in group
+	for _, g := range groups {
+		g := g
+		base := groupBaseIndex
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			var err error
+			if g[0].Kind == driver.Get {
+				err = c.runGets(ctx, g, opts)
+			} else {
+				err = c.runWrites(ctx, g, opts)
+			}
+			for i := 0; i < len(g); i++ {
+				errs[base+i] = err
+			}
+		}()
+		groupBaseIndex += len(g)
+	}
+	wg.Wait()
+	return driver.NewActionListError(errs)
+}
+
+// splitActionsUnordered divides the actions slice into sub-slices, which
+// maximizes the actions sent in each group and may change the order of the
+// action list, each of which can be passed to run a dynamo transaction
+// operation.
+func (c *collection) splitActionsUnordered(actions []*driver.Action) ([][]*driver.Action, error) {
+	var (
+		groups [][]*driver.Action              // the actions, split; the return value
+		gets   []*driver.Action                // the gets group currently being constructed
+		writes []*driver.Action                // the writes group currently being constructed
+		km     = make(map[[2]interface{}]bool) // keys should be unique across all groups in unordered mode
+	)
+	collect := func(cur *[]*driver.Action) { // called when the current group is known to be finished
+		if len(*cur) > 0 {
+			groups = append(groups, *cur)
+			*cur = nil
+		}
+	}
+	for _, a := range actions {
+		if keys := c.primaryKey(a); keys[0] != nil {
+			if km[keys] {
+				return nil, fmt.Errorf("repeated key: %v, %v; item must be unique for unordered actions", keys[0], keys[1])
```

Comment:
You can return the index of the bad action (the second one) as well.

Reply:
Done.